[json-spirit] Add initial port

smlee-hdactech opened the pull request:
I added the porting job for json-spirit.
Have a nice day.

Commit:
[json-spirit] Add initial port

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5157) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5157) before we can accept your contribution.<br/><br/>:x: smlee-hdactech [sign now](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5157)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vcpkg?pullRequest=5157) it.</sub>

Codiferous commented on ports/json-spirit/portfile.cmake:
```diff
@@ -0,0 +1,49 @@
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO smlee-hdactech/json_spirit
```
This appears to be a fork of the library. We try to use the "official" upstream for every package in vcpkg unless there's an excellent reason not to. 

It seems that the most appropriate upstream to use for this is https://github.com/png85/json_spirit, based on what system package managers like Homebrew and Debian use.

Could you update this PR to use a more appropriate upstream?

smlee-hdactech replied:
O.K. I got it. I will check it using original stream.

smlee-hdactech replied:
I did pull-request to original stream, png85, however, the repository maybe updates no longer.
How about my creating new repository for json-spirit?

Codiferous replied:
We try to use the upstream repo as much as possible with vcpkg. I've made the changes to this PR to use the png85 repo and modified the PR to only allow static linking. When/if the upstream repo accepts your PR, we can modify our port to reflect the DLL use case.

Codiferous commented on ports/json-spirit/portfile.cmake:
```diff
@@ -0,0 +1,49 @@
+    HEAD_REF master
+)
+
+#set(SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/json_spirit-4.1.0)
```
I believe these comments were not intentionally left here. Could you remove them when you update the upstream?

smlee-hdactech replied:
Of course, I will remove it.

Codiferous committed:
Modify PR to using pn85, enable some features

json-spirit is a weird case because the main repo that package managers
tend to use appears to not be actively maintained. Nevertheless, vcpkg
tries to use the "official" source of a repository if at all possible.

In order to do this, I've made the following changes to the PR:
- I've changed the upstream REPO, REF, and SHA512.
- I've passed options to disable building the demos and tests
- I've made the project build only as a static library.

The reason for this is that vcpkg does not support dynamical linkin.
Another reason is that the library as it is in the repo does not have
all the definitions correctly annotated.smlee-hdactech has made a PR
against the major upstream repo to fix this, butwe are going to defer
to the main repo until that change is accepted.

Codiferous commented:
I've modified this PR to pull from the png85 repo because that is the source that other repositories use. This means that until your PR containing changes to enable dllexport on some of the entities is accepted in png85, vcpkg will only allow static linkage for json-spirit. 

If your PR is not accepted due to the mainstream repo being inactive, please submit another PR to vcpkg that contains patch files to apply to the json-spirit source. 

Thanks!
